Initial Support for Spark 4.0 preview

huaxingao opened the pull request.

huaxingao commented on .github/workflows/java-ci.yml:
```diff
@@ -95,7 +95,7 @@ jobs:
     strategy:
       matrix:
-        jvm: [11, 17, 21]
+        jvm: [17, 21]
```
This is for `build-checks`. Not able to do `build-checks` using java 11 because Spark 4.0 is build using java 17.

huaxingao commented on .github/workflows/java-ci.yml:
```diff
@@ -108,7 +108,7 @@ jobs:
     strategy:
       matrix:
-        jvm: [11, 17, 21]
+        jvm: [17, 21]
```
This is for `build-javadoc`. Not able to do `build-javadoc` using java 11 because Spark 4.0 is build using java 17.

RussellSpitzer replied:
Is there a reason to do build-javadoc on more than one JVM anyway?

huaxingao commented:
CI for preview1 passed. 
CI for preview2 failed.
Trying SNAPSHOT to see if some of the Spark issues in preview2 have been fixed in SNAPSHOT.

RussellSpitzer commented on gradle/libs.versions.toml:
```diff
@@ -47,6 +47,7 @@ flink120 = { strictly = "1.20.0"}
 google-libraries-bom = "26.47.0"
 guava = "33.3.0-jre"
 hadoop2 = "2.7.3"
+hadoop34 = "3.4.0" # For Spark 4.0 support
```
tiniest nit: This goes after hadoop3

huaxingao replied:
Changed. Thanks

RussellSpitzer commented on gradle/libs.versions.toml:
```diff
@@ -137,6 +139,7 @@ hadoop2-common = { module = "org.apache.hadoop:hadoop-common", version.ref = "ha
 hadoop2-hdfs = { module = "org.apache.hadoop:hadoop-hdfs", version.ref = "hadoop2" }
 hadoop2-mapreduce-client-core = { module = "org.apache.hadoop:hadoop-mapreduce-client-core", version.ref = "hadoop2" }
 hadoop2-minicluster = { module = "org.apache.hadoop:hadoop-minicluster", version.ref = "hadoop2" }
+hadoop34-minicluster = { module = "org.apache.hadoop:hadoop-minicluster", version.ref = "hadoop34" }
```
after hadoop3

huaxingao replied:
Changed

RussellSpitzer commented on spark/v4.0/build.gradle:
```diff
@@ -33,6 +34,7 @@ configure(sparkProjects) {
         force "com.fasterxml.jackson.module:jackson-module-scala_${scalaVersion}:${libs.versions.jackson215.get()}"
         force "com.fasterxml.jackson.core:jackson-databind:${libs.versions.jackson215.get()}"
         force "com.fasterxml.jackson.core:jackson-core:${libs.versions.jackson215.get()}"
+        force "org.antlr:antlr4:${antlrVersion}"
```
Why do we have to force here? Shouldn't we be ok just pulling the Antlr out of the spark version?

huaxingao replied:
Shouldn't need `force`. Fixed. Thanks

RussellSpitzer commented on spark/v4.0/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala:
```diff
@@ -30,7 +30,7 @@ import org.apache.iceberg.NullOrder
 import org.apache.iceberg.expressions.Term
 import org.apache.iceberg.spark.Spark3Util
-import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.analysis.IcebergAnalysisException
```
Why are we switching to our internal Exception class here?

huaxingao replied:
Because Spark 4.0 no longer allows the construction of a new `AnalysisException` with just a string message. I actually have a separate [PR](https://github.com/apache/iceberg/pull/10766) for this. We can probably merge this change in Spark 3.5, although it's not strictly required there.

RussellSpitzer commented on spark/v4.0/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestSystemFunctionPushDownInRowLevelOperations.java:
```diff
@@ -260,7 +267,12 @@ private void checkUpdate(RowLevelOperationMode mode, String cond) {
               DistributionMode.NONE.modeName());
 
           Dataset<Row> changeDF = spark.table(tableName).where(cond).limit(2).select("id");
-          changeDF.coalesce(1).writeTo(tableName(CHANGES_TABLE_NAME)).create();
```
Why are we rethrowing here?

huaxingao replied:
I added the catch block to make the code compile for preview2. Preview1 doesn't need this. After switching to Preview2, I actually got a bunch of CI failures due to `TableAlreadyExistsException` in a few test suites. Preview1 works fine. I am still trying to figure out which change between Preview1 and Preview2 caused the behavior changes for `TableAlreadyExistsException`.

RussellSpitzer commented on spark/v4.0/spark/src/main/scala/org/apache/spark/sql/stats/ThetaSketchAgg.scala:
```diff
@@ -119,3 +122,12 @@ case class ThetaSketchAgg(
+  def expr(node: ColumnNode): Expression = {
+    node match {
+      case ExpressionColumnNode(expression, _) => expression
+      case node => throw SparkException.internalError("Unsupported ColumnNode: " + node)
```
Should we be throwing a Spark Internal error here? Seems like this our issue?

huaxingao replied:
I will change this to an Iceberg Exception. I am not making the change in this round because I want to try Preview 1 to see if the other changes can pass the CI. I will fix this later when I try Preview 2.

RussellSpitzer commented on spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteDataFilesAction.java:
```diff
@@ -138,6 +138,7 @@ public class TestRewriteDataFilesAction extends TestBase {
   public static void setupSpark() {
     // disable AQE as tests assume that writes generate a particular number of files
     spark.conf().set(SQLConf.ADAPTIVE_EXECUTION_ENABLED().key(), "false");
+    spark.conf().set(SQLConf.ANSI_ENABLED().key(), "false");
```
Why is this needed?

huaxingao replied:
I actually meant to change only the ANSI setting for `TestRewriteDataFilesAction#testZOrderAllTypesSort`, because CAST(longCol AS BINARY) is not allowed with ANSI, which is the default now for Spark 4.0. 
I fixed the test to use ANSI  false for testZOrderAllTypesSort.

RussellSpitzer commented on spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java:
```diff
@@ -108,14 +108,14 @@ public static Object[][] parameters() {
         SparkCatalogConfig.SPARK.implementation(),
         SparkCatalogConfig.SPARK.properties(),
         PARQUET,
-        ImmutableMap.of(COMPRESSION_CODEC, "zstd", COMPRESSION_LEVEL, "1")
```
nit: We are only partially alphabetizing here

This is the kind of thing I do love but it should be

gzip
snappy
zstd

I would probably just skip this change for now and do it in another PR

huaxingao replied:
The reason I wanted to `ImmutableMap.of(COMPRESSION_CODEC, "zstd", COMPRESSION_LEVEL, "1")` after gzip is that the new Hadoop version uses `CompressionLevel` to initialize a GzipCompressor, and this COMPRESSION_LEVEL, "1", is carried over to gzip. However, "1" is not a valid compression level for gzip, so it throws an exception.
```
    org.apache.spark.SparkException: Job aborted due to stage failure: Task 1 in stage 6.0 failed 1 times, most recent failure: Lost task 1.0 in stage 6.0 (TID 7) (192.168.50.141 executor driver): java.lang.IllegalArgumentException: No enum constant org.apache.hadoop.io.compress.zlib.ZlibCompressor.CompressionLevel.1
    	at java.base/java.lang.Enum.valueOf(Enum.java:273)
    	at org.apache.hadoop.conf.Configuration.getEnum(Configuration.java:1786)
    	at org.apache.hadoop.io.compress.zlib.ZlibFactory.getCompressionLevel(ZlibFactory.java:165)
    	at org.apache.hadoop.io.compress.zlib.BuiltInGzipCompressor.init(BuiltInGzipCompressor.java:157)
    	at org.apache.hadoop.io.compress.zlib.BuiltInGzipCompressor.<init>(BuiltInGzipCompressor.java:67)
    	at org.apache.hadoop.io.compress.GzipCodec.createCompressor(GzipCodec.java:64)
    	at org.apache.hadoop.io.compress.CodecPool.getCompressor(CodecPool.java:152)
    	at org.apache.hadoop.io.compress.CodecPool.getCompressor(CodecPool.java:168)
    	at org.apache.parquet.hadoop.CodecFactory$HeapBytesCompressor.<init>(CodecFactory.java:157)
    	at org.apache.parquet.hadoop.CodecFactory.createCompressor(CodecFactory.java:219)
    	at org.apache.parquet.hadoop.CodecFactory.getCompressor(CodecFactory.java:202)
    	at org.apache.iceberg.parquet.ParquetWriter.<init>(ParquetWriter.java:90)
    	at org.apache.iceberg.parquet.Parquet$WriteBuilder.build(Parquet.java:360)
```
I think this over; rather than switching the order, it's better to unset the COMPRESSION_CODEC and COMPRESSION_LEVEL for each test.

huaxingao replied:
I probably should have a separate PR and fix this in Spark3.5 too. WDYT?

RussellSpitzer commented on spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/source/TestStructuredStreaming.java:
```diff
@@ -302,4 +303,12 @@ private <T> MemoryStream<T> newMemoryStream(int id, SQLContext sqlContext, Encod
   }
+
+  private void deleteFileAndCrc(File file) throws IOException {
+    File crcFile = new File(file.getParent(), "." + file.getName() + ".crc");
```
Are you sure the "." +file.getName is correct here?

Could we use 
```new File(file.getAbsolutePath + ".crc")```

Also under what cases should crcFile not exist?

huaxingao replied:
file.getAbsolutePath + ".crc" has
```
/var/folders/lp/nhx1yjj90qb54xq7v_grpl9c0000gn/T/junit3203001525697015981/parquet/checkpoint/commits/1.crc
```
but I need 
```
/var/folders/lp/nhx1yjj90qb54xq7v_grpl9c0000gn/T/junit3203001525697015981/parquet/checkpoint/commits/.1.crc
```

When deleting the commits like 
```
      // remove the last commit to force Spark to reprocess batch #1
      File lastCommitFile = new File(checkpoint.toString() + "/commits/1");
      assertThat(lastCommitFile.delete()).as("The commit file must be deleted").isTrue();
```
I got Exception so I manually delete the crc file
```
 Caused by:
            org.apache.hadoop.fs.FileAlreadyExistsException: Rename destination file:/var/folders/lp/nhx1yjj90qb54xq7v_grpl9c0000gn/T/junit16535101756839742517/parquet/checkpoint/commits/.1.crc already exists.
                at app//org.apache.hadoop.fs.AbstractFileSystem.renameInternal(AbstractFileSystem.java:876)
                at app//org.apache.hadoop.fs.AbstractFileSystem.rename(AbstractFileSystem.java:807)
                at app//org.apache.hadoop.fs.ChecksumFs.renameInternal(ChecksumFs.java:519)
                at app//org.apache.hadoop.fs.AbstractFileSystem.rename(AbstractFileSystem.java:807)
                at app//org.apache.hadoop.fs.FileContext.rename(FileContext.java:1044)
                at app//org.apache.spark.sql.execution.streaming.FileContextBasedCheckpointFileManager.renameTempFile(CheckpointFileManager.scala:376)
                at app//org.apache.spark.sql.execution.streaming.CheckpointFileManager$RenameBasedFSDataOutputStream.close(CheckpointFileManager.scala:156)
                at app//org.apache.spark.sql.execution.streaming.HDFSMetadataLog.write(HDFSMetadataLog.scala:207)
```

huaxingao commented:
@RussellSpitzer Thanks for your review! I have addressed the comments and switched back to Preview1, along with reverting a few changes I made for Preview2/snapshot. I switched back to Preview1 to test if my changes can pass the CI, since I haven't made Preview2 work yet. Could you please take another look when you have time? Thanks!

RussellSpitzer commented:
I'm ready for V2 :) Let me know when you have those changes up. I'm trying to review this on a per commit basis because the diff is so large :)
